csharp/test/Microsoft.ML.OnnxRuntime.Tests.NetCoreApp/TrainingTest.cs

```diff
@@ -0,0 +1,287 @@
+                {
+                    var memInfo = OrtMemoryInfo.DefaultInstance; // CPU
+
+                    // Create inputs
+                    long[] inputShape = { 2, 784 };
+                    var byteSize = inputShape.Aggregate(1L, (a, b) => a * b) * sizeof(float);
+                    pinnedInputs.Add(FixedBufferOnnxValue.CreateFromMemory<float>(memInfo, input,
+                        TensorElementType.Float, inputShape, byteSize));
+
+                    long[] labelsShape = { 2 };
+                    byteSize = labelsShape.Aggregate(1L, (a, b) => a * b) * sizeof(Int32);
+                    pinnedInputs.Add(FixedBufferOnnxValue.CreateFromMemory<Int32>(memInfo, labels,
+                        TensorElementType.Int32, labelsShape, byteSize));
+
+
+                    // Prepare output buffer
+                    long[] outputShape = { };
+                    byteSize = outputShape.Aggregate(1L, (a, b) => a * b) * sizeof(float);
+                    float[] outputBuffer = new float[expectedOutput_1.Length];
+                    pinnedOutputs.Add(FixedBufferOnnxValue.CreateFromMemory<float>(memInfo, outputBuffer,
+                        TensorElementType.Float, outputShape, byteSize));
+
+                    trainingSession.TrainStep(pinnedInputs, pinnedOutputs);
+                    Assert.Equal(expectedOutput_1, outputBuffer, new FloatComparer());
+
+                    trainingSession.ResetGrad();
+
+                    trainingSession.TrainStep(pinnedInputs, pinnedOutputs);
+                    Assert.Equal(expectedOutput_1, outputBuffer, new FloatComparer());
+
+                    trainingSession.OptimizerStep();
+
+                    trainingSession.TrainStep(pinnedInputs, pinnedOutputs);
+                    Assert.Equal(expectedOutput_2, outputBuffer, new FloatComparer());
+                }
+
+                trainingSession.Dispose();
+            }
+        }
+
+        [Fact(DisplayName = "TestTrainingSessionSaveCheckpoint")]
+        public void TestTrainingSessionSaveCheckpoint()
+        {
+            using (var state = new CheckpointState())
+            {
+                Assert.NotNull(state);
+                string checkpointPath = Path.Combine(Directory.GetCurrentDirectory(), "checkpoint.ckpt");
+                state.LoadCheckpoint(checkpointPath);
+                string trainingPath = Path.Combine(Directory.GetCurrentDirectory(), "training_model.onnx");
+
+                var trainingSession = new TrainingSession(state, trainingPath);
+
+                string savedCheckpointPath = Path.Combine(Directory.GetCurrentDirectory(), "saved_checkpoint.ckpt");
+                trainingSession.SaveCheckpoint(savedCheckpointPath);
+                trainingSession.Dispose();
+
+                var loadedState = new CheckpointState();
+                loadedState.LoadCheckpoint(savedCheckpointPath);
+                var newTrainingSession = new TrainingSession(loadedState, trainingPath);
```

Comment:
These are not being dispose of on exception. Must be used with either using clause or added to a disposable container.